components/scream/src/diagnostics/atm_density.hpp

```diff
@@ -0,0 +1,104 @@
+#ifndef EAMXX_ATM_DENSITY_DIAGNOSTIC_HPP
+#define EAMXX_ATM_DENSITY_DIAGNOSTIC_HPP
+
+#include "share/atm_process/atmosphere_diagnostic.hpp"
+#include "share/util/scream_common_physics_functions.hpp"
+
+namespace scream
+{
+
+/*
+ * This diagnostic will produce the potential temperature.
+ */
+
+class AtmDensityDiagnostic : public AtmosphereDiagnostic
+{
+public:
+  template <typename S>
+  using SmallPack     = ekat::Pack<S,SCREAM_SMALL_PACK_SIZE>;
+
+  using Spack         = SmallPack<Real>;
+  using Pack          = ekat::Pack<Real,Spack::n>;
+  using PF            = scream::PhysicsFunctions<DefaultDevice>;
+  using KT            = KokkosTypes<DefaultDevice>;
+  using view_2d       = typename KT::template view_2d<Spack>;
+  using view_2d_const = typename KT::template view_2d<const Spack>;
+
+  // Constructors
+  AtmDensityDiagnostic (const ekat::Comm& comm, const ekat::ParameterList& params);
+
+  // Set type to diagnostic
+  AtmosphereProcessType type () const { return AtmosphereProcessType::Diagnostic; }
+
+  // The name of the diagnostic
+  std::string name () const { return "Atmosphere Density"; } 
+
+  // Get the required grid for the diagnostic
+  std::set<std::string> get_required_grids () const {
+    static std::set<std::string> s;
+    s.insert(m_params.get<std::string>("Grid"));
+    return s;
+  }
+
+  // Set the grid
+  void set_grids (const std::shared_ptr<const GridsManager> grids_manager);
+
+  // Actual diagnostic calculation 
+  struct run_diagnostic_impl {
```

Comment:
This regards all the diagnostics in the PR.

Is this inner functor class necessary? It seems to add an unnecessary layer of complexity. Can't you just have the operator() in the AtmDensityDiagnostic class?